### ISSUE TITLE
Show Filters - "User" Apps Appeared Twice

### ISSUE THREAD

**Issue by 0pLuS0** [ukanth/afwall#871]
My Specs:
--------------
OnePlus 5T
Oxygen OS 5.1.5 - Oreo 8.1.0
Magisk 17.2
AFWall+ 2.9.9 (From F-Droid)
Xposed 3.1.5 - AFwall+ Module
--------------

I wish I had more to go on, I wasn't really paying attention to what exactly triggered this. I went back and have been playing with the various settings, but I have not been able to replicate it.

When I first enabled the "Show filter" and one point, when I had it on "Users" all the apps appeared twice in the menu.

Hopefully the logcat I have attached will give you information over the problem, as to why they appeared twice.

Keep up the great work, thanks! :+1: 

[logcat.txt](https://github.com/ukanth/afwall/files/2433463/logcat.txt)

**Comment by 0pLuS0** [comment]
Ok this happened again, this time I took a screen shot.

All I did was open AFWall+ click on the User profile, and the apps appeared twice.

I grabbed a logcat the minute it happened, so hopefully it will be of some help.


![afwall_2x](https://user-images.githubusercontent.com/36190223/46334448-99b11e00-c613-11e8-8285-a32f02446e1c.jpg)


```LOGCAT```

[logcat.txt](https://github.com/ukanth/afwall/files/2436833/logcat.txt)

**Comment by 0lm** [comment]
if you have multiple profiles on your phone, afwall will show the apps double, if they are also cloned/installed there. did you activate dual app support and multiuser support in experimental, by any chance?  i know that behavior thanks to the shelter app. when i clone apps to the shelter, they appear twice in afwall, when they are existing in both  profiles.

**Comment by 0pLuS0** [comment]
This was done on a fresh clean install then deodex, patched sig spoof, debloated and micrgo and no multipile profiles only single, nothing cloned installed...

Dual app support and multiuser is not active, I did not enable this.

**Comment by 0lm** [comment]
then i have no idea. maybe it has something to do with oxygen. but unfortunately i cant say anything to that, because im using lineageOS.

**Comment by 0pLuS0** [comment]
Logcat doesn't show anything?

Do you have a Debug version of AFWall?

**Comment by 0lm** [comment]
im not good at reading logcats.  thats why i cant say anythong to the logcat. i tried to figure out the problem with known problems.  no i have the normal version of afwall from yalpstore. if you freshly installed it.. did you try to wipe the rom and reinstall the rom? maybe there really went something wrong during installination

EDIT: i found out something interesting. according to Oxygen 5.1.5, there should be a parralel apps feature (idk if it was in the beta or already stable released). did you activate parralel apps by any chance? doing that, all important system apps will be cloned so they can be used in another profile.  its a similar behaviour to shelter.

**Comment by 0pLuS0** [comment]
It was a clean install and no parallel apps...

**Comment by ukanth** [comment]
I'm not able to produce this. Logs won't help in this.

**Comment by 0pLuS0** [comment]
@ukanth Hi...

I'm hacking away doing other things, so when I get a moment I'll reinstall AFWall on a clean install and see how it goes again, hopefully it was just something wrong before and won't happen again.

I'll keep you posted!

Thanks :)

**Comment by 0pLuS0** [comment]
@ukanth Hi...

I mentioned a ```DEBUG``` version before, but you didn't say anything, do you have a Debug version? Because it seems, as you mentioned logcat doesn't help you, so we at least need this to track down bugs... yes/no? Otherwise what else is there, if not logcat and a debug version...

I'm hacking away doing other things, so when I get a moment I'll reinstall AFWall on a clean install and see how it goes again, hopefully it was just something wrong before and won't happen again.

I'll keep you posted!

Thanks :)

**Comment by ukanth** [comment]
debug means you compiled yourself ?

**Comment by 0pLuS0** [comment]
@ukanth sorry, yes of course, but I don't have the means at the moment to compile it on my end, otherwise I'd do this. 

What I meant to ask, my mistake, is if you could please provide a Debug version on the Releases section here on Github, for end-users to download.

Of course for me for the moment, but it would be helpful to do so for future releases....

https://github.com/ukanth/afwall/releases

Thanks

**Comment by ukanth** [comment]
Not all users willing/spend time to get logs and provide to developers in terms of crashes. I get crash logs from playstore (not much meaningful ones, but enough to identify and provide fix).

Releasing multiple flavours leads to confusion in my opinion.

**Comment by 0pLuS0** [comment]
@ukanth yes, I understand, but the Geeks are the ones that typically find their way to GitHub. You can simply put text in the Releases section that explains to not download the Debug version, unless you are using it to report/find bugs...

Easy enough... :)

In the mean time, if you can build a Debug version for me, then if I run into this problem again, hopefully it will help.

But one last thing I forgot to mention, would I then need to run ```gdb``` on the phone for output? Or is there a way this can be self contained in the apk, so any bugs/crashes that come up, it can self-create the logs it needs?

**Comment by 0pLuS0** [comment]
I wonder if ```Show Filters``` is the problem because with it disabled I'm not seeing the issue, at least I haven't seen it appear yet...

Also with the ```Show Filters``` disabled all the apps are staying in alphabetical order, which was another issue I saw with the ```Show Filters``` enabled...

hmm

**Comment by HerborenInVlammen** [comment]
I am having exactly the same issue on my Xiomi Redmi Note 4
Slimrom Nougat 7.1.2
Rooted TWRP SuperSU


This happens randomly, and  directly after installing a new app, and using the notification to set Internet permissions.  Afwall set to Filter by USER / Sort By Install Date.

Have to Force Stop or reboot to get it show properly again.  Started with new version.   It is definitely an issue as more than one user has the problem, and on totally different configs.

To reproduce set Filter and Sort like mentioned above, and start installing some apps while using the Notication to Set Internet permissions.  It will occur sooner or later.

**Comment by 0pLuS0** [comment]
@ukanth 

This has happened again with OxygenOS 5.1.7 and AFWall+ 3.0.1

This time I believe I have figured out possibly what triggers this.

Everything was find until I downloaded from the Play Store the ```EX Kernel Manager```

The first thing I noticed after installing it and then looking in AFWall+, was that EX Kernel Manager was at the top of all the other apps, I assumed because ```EX``` was in UPPER CASE letters, casing it to appear at the top.

I check marked EX Kernel Manager for wifi for the ```Allow selected``` section, then I closed and reopened AFWall+, still noticing EX Kernel Manager at the top of the list. So I unchecked it and closed and reopened AFWall+ and this time it was not at the top of the list, it had been alphabetically arranged, but now I have DOUBLE ENTRY of some applications as before... :(

```So maybe some words like this in UPPER CASE or something else odd in the names, is causing this?```

![afwall](https://user-images.githubusercontent.com/36190223/49059073-fa696b00-f1fe-11e8-8f83-42ed816d0fed.jpg)

**Comment by 0pLuS0** [comment]
@ukanth 

This has happened again with OxygenOS 5.1.7 and AFWall+ 3.0.1

This time I believe I have figured out possibly what triggers this.

Everything was fine until I downloaded from the Play Store the ```EX Kernel Manager```

The first thing I noticed after installing it and then looking in AFWall+, was that EX Kernel Manager was at the top of all the other apps, I assumed because ```EX``` was in UPPER CASE letters, casing it to appear at the top.

I check marked EX Kernel Manager for wifi, for the ```Allow selected``` section, then I closed and reopened AFWall+, still noticing EX Kernel Manager at the top of the list. So I unchecked it and closed and reopened AFWall+ and this time it was not at the top of the list, it had been alphabetically arranged, but now I noticed DOUBLE ENTRY of some applications as before... :(

```So maybe some words like this in UPPER CASE or something else odd in names, is causing this because before this happened it was ok.```

![afwall](https://user-images.githubusercontent.com/36190223/49059073-fa696b00-f1fe-11e8-8f83-42ed816d0fed.jpg)

**Comment by 0pLuS0** [comment]
@ukanth 

This has happened again with OxygenOS 5.1.7 and AFWall+ 3.0.1

This time I believe I have figured out possibly what triggers this.

Everything was fine until I downloaded from the Play Store the ```EX Kernel Manager```

The first thing I noticed after installing it and then looking in AFWall+, was that EX Kernel Manager was at the top of all the other apps, I assumed because ```EX``` was in UPPER CASE letters, casing it to appear at the top.

I check marked EX Kernel Manager for wifi, for the ```Allow selected``` section, then I closed and reopened AFWall+, still noticing EX Kernel Manager at the top of the list. So I unchecked it and closed and reopened AFWall+ and this time it was not at the top of the list, it had been alphabetically arranged, but now I noticed DOUBLE ENTRY of some applications as before... :(

```So maybe some words like this in UPPER CASE or something else odd in names, is causing this because before this happened it was ok.```

```I cleared the phones cache, and restarted the phone, and now it's all back to normal```

![afwall](https://user-images.githubusercontent.com/36190223/49059073-fa696b00-f1fe-11e8-8f83-42ed816d0fed.jpg)

**Comment by 0pLuS0** [comment]
Well I thought after clearing the cache and rebooting it would stay ok, but it returned again.

This is really bothersome seeing all these double entries, I hope you can PLEASE figure this out.

Now Aurora Store appears twice, but in the last screen shot you'll notice it's only once, odd that another app gets added double... :face_with_head_bandage: 


![double_apps](https://user-images.githubusercontent.com/36190223/49060967-913a2580-f207-11e8-82c4-055bfd39b655.jpg)

**Comment by 0pLuS0** [comment]
@ukanth 

So the problem returned again, and a simple wipe of the cache didn't do the trick this time. So now I undid all my allowed apps, put the preferences back, disabled the firewall and cleared the cache and data and rebooted, and then set it all up over again, and now it's back to normal.

But really some effort to get it back, having to wipe it all out and start over, I hope you can PLEASE figure this out.

Now Aurora Store appears twice, but in the last screen shot you'll notice it's only once, odd that another app gets added double... :face_with_head_bandage: 


![double_apps](https://user-images.githubusercontent.com/36190223/49060967-913a2580-f207-11e8-82c4-055bfd39b655.jpg)

**Comment by 0pLuS0** [comment]
@ukanth so can you make a DEBUG version that end-users can download off of GitHub to test for bugs like this?

I'm assuming a DEBUG version will help you track down bugs? 

Thanks

**Comment by Abu-Khulzat** [comment]
Exact same thing happened on Oneplus 6t OS 9.0.7. Now sure what caused this but after clearing data for AFWall all is OK

**Comment by ukanth** [comment]
Can you please check this version ? Problem is I'm unable to reproduce this behavior on all my devices.

https://drive.google.com/open?id=1dYgchCqcsjENdOjYjVrnnwWtxtV94bav

**Comment by Abu-Khulzat** [comment]
I don't see double apps so far, but the app is not able to enable the firewall. When I press enebale firewall, I get a message "error purging tables". And in the notification says the firewall is disabled. I tried to clear cash, data.

**Comment by ukanth** [comment]
You have flush the rules. Menu -show rules - purge rules

**Comment by HerborenInVlammen** [comment]
Update to fix the duplicate display issue does not work.  AAMOF it is significantly worse.

Note:  This also happens on my old ASUS Zenfone 2, and on my son's Oneplus 6.

**Comment by HerborenInVlammen** [comment]
Note:  I can almost repeat this at will with the latest update.  Almost but not quite.

Steps:
1) Sort by user app.
2) Force-close Afwall+
3) Restart Afwall+
8 of 10 times it starts with duplicated display of user apps.

**Comment by DogDay7** [comment]
SOLUTION: I had the same problem too. I don't know what causes this problem but I found a really easy fix to remove the multiple listing of same apps on the firewall permission filter page. All u need to do is go to top of list and then swipe down on the screen. This refreshes the page. A pop-up says 'Retrieving installed apps". That removed the multiple listing if apps fast and easy. For now this is the solution I use. Hope this helps. Good Luck !
